### PR TITLE
TAJO-751 JDBC driver should support cancel() method

### PR DESCRIPTION
**Pull request by navis**
Not yet added test cases

### CODE REVIEW

**Comment by blrunner** [comment]
@navis 

Thanks for your contribution. But I failed to run unit test. For the reference, 51 cases were failures and 3 cases were errors. I attached some log messages as following:

```
Tests run: 3, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 2.248 sec <<< FAILURE! - in org.apache.tajo.querymaster.TestTaskStatusUpdate
case3(org.apache.tajo.querymaster.TestTaskStatusUpdate)  Time elapsed: 1.093 sec  <<< FAILURE!
java.lang.AssertionError: expected:<4> but was:<2>
        at org.junit.Assert.fail(Assert.java:88)
        at org.junit.Assert.failNotEquals(Assert.java:743)
        at org.junit.Assert.assertEquals(Assert.java:118)
        at org.junit.Assert.assertEquals(Assert.java:555)
        at org.junit.Assert.assertEquals(Assert.java:542)
        at org.apache.tajo.querymaster.TestTaskStatusUpdate.assertStatus(TestTaskStatusUpdate.java:164)
        at org.apache.tajo.querymaster.TestTaskStatusUpdate.case3(TestTaskStatusUpdate.java:111)
```

```
testInsertOverwriteIntoParquet(org.apache.tajo.engine.query.TestInsertQuery)  Time elapsed: 0.002 sec  <<< ERROR!
com.google.protobuf.ServiceException: relation "TestInsertQuery.parquet_table" already exists

        at org.apache.tajo.client.QueryClientImpl.executeQueryAndGetResult(QueryClientImpl.java:210)
        at org.apache.tajo.client.TajoClientImpl.executeQueryAndGetResult(TajoClientImpl.java:110)
        at org.apache.tajo.QueryTestCaseBase.executeString(QueryTestCaseBase.java:323)
        at org.apache.tajo.engine.query.TestInsertQuery.testInsertOverwriteIntoParquet(TestInsertQuery.java:546)
```

**Comment by navis** [comment]
Fixed test fail

**Comment by blrunner** [comment]
Thanks @navis 
I'll review today. :)

**Comment by navis** [comment]
Addressed comment by Jungryong Lee.

**Comment by blrunner** [comment]
@navis 

Could you rebase this? I failed to merge the patch as following:
```
Auto-merging tajo-jdbc/src/main/java/org/apache/tajo/jdbc/TajoStatement.java
CONFLICT (content): Merge conflict in tajo-jdbc/src/main/java/org/apache/tajo/jdbc/TajoStatement.java
Auto-merging tajo-jdbc/src/main/java/org/apache/tajo/jdbc/TajoPreparedStatement.java
Auto-merging tajo-client/src/main/java/org/apache/tajo/client/SessionConnection.java
Automatic merge failed; fix conflicts and then commit the result.
```

**Comment by navis** [comment]
rebased to master

**Comment by blrunner** [comment]
@navis 

Could you check it again?
I still fail to merge this.

**Comment by navis** [comment]
Sorry, I've missed recent update of master.

**Comment by blrunner** [comment]
No problem. :)

**Comment by blrunner** [comment]
@navis 

Thanks for your contribution.
Could you add some unit test cases for this issue?

**Comment by navis** [comment]
Added a testcase & minor refactorings

**Comment by navis** [comment]
@blrunner Could you review this?

**Comment by blrunner** [comment]
@navis 

Sorry for late review.
I'll review this again soon. :)

**Comment by blrunner** [comment]
Hi @navis 

I failed to build this as follows.

```
-------------------------------------------------------------------------------
Test set: org.apache.tajo.cli.tsql.TestTajoCli
-------------------------------------------------------------------------------
Tests run: 18, Failures: 1, Errors: 0, Skipped: 0, Time elapsed: 5.29 sec <<< FAILURE! - in org.apache.tajo.cli.tsql.TestTajoCli
testHelpSessionVars(org.apache.tajo.cli.tsql.TestTajoCli)  Time elapsed: 0.01 sec  <<< FAILURE!
org.junit.ComparisonFailure: expected:<...me from Master
\set [FETCH_BLOCK] [true or false] - W...> but was:<...me from Master
\set [BLOCK_ON_RESULT] [true or false] - W...>
        at org.junit.Assert.assertEquals(Assert.java:115)
        at org.junit.Assert.assertEquals(Assert.java:144)
        at org.apache.tajo.cli.tsql.TestTajoCli.assertOutputResult(TestTajoCli.java:117)
        at org.apache.tajo.cli.tsql.TestTajoCli.assertOutputResult(TestTajoCli.java:99)
        at org.apache.tajo.cli.tsql.TestTajoCli.assertOutputResult(TestTajoCli.java:95)
        at org.apache.tajo.cli.tsql.TestTajoCli.testHelpSessionVars(TestTajoCli.java:366)
```

Could you check above error?

**Comment by navis** [comment]
seemed needed to be rebased. I'll do that

**Comment by blrunner** [comment]
Okay, thanks. :)

**Comment by navis** [comment]
done.

**Comment by blrunner** [comment]
Thanks @navis 
I'll review this today. :-)

File: tajo-client/src/main/java/org/apache/tajo/jdbc/WaitingResultSet.java [hunk 1]
```diff
@@ -0,0 +1,71 @@
...
+
+  @Override
+  public boolean next() throws SQLException {
+    getSchema();
+    return super.next();
+  }
+
+  @Override
+  protected Schema getSchema() throws SQLException {
+    return schema == null ? schema = waitOnResult() : schema;
```
Comment (blrunner) [inline tajo-client/src/main/java/org/apache/tajo/jdbc/WaitingResultSet.java]: Could you explain why you check schema?

File: tajo-core/src/main/java/org/apache/tajo/master/TajoMasterClientService.java [hunk 2]
```diff
@@ -623,6 +623,12 @@ public BoolProto killQuery(RpcController controller, QueryIdRequest request) thr
       try {
         context.getSessionManager().touch(request.getSessionId().getId());
         QueryId queryId = new QueryId(request.getQueryId());
+
+        QueryInProgress progress = context.getQueryJobManager().getQueryInProgress(queryId);
+        if (progress == null || progress.isFinishState() || progress.isFinishing()) {
```
Comment (blrunner) [inline tajo-core/src/main/java/org/apache/tajo/master/TajoMasterClientService.java]: Do we need to check the kill wait status surely?

**Comment by blrunner** [comment]
Hi @navis 

This looks good overall to me and I leaved some comments. 
In addition, if possible, could rebase this again? I failed to merge as follows.

```
Removing tajo-rpc/tajo-rpc-protobuf/src/main/java/org/apache/tajo/rpc/RpcConnectionPool.java
Auto-merging tajo-core/src/main/java/org/apache/tajo/master/QueryInProgress.java
Auto-merging tajo-client/src/main/java/org/apache/tajo/client/SessionConnection.java
CONFLICT (content): Merge conflict in tajo-client/src/main/java/org/apache/tajo/client/SessionConnection.java
Auto-merging tajo-client/src/main/java/org/apache/tajo/client/QueryClientImpl.java
Automatic merge failed; fix conflicts and then commit the result.
```

**Pull request closed without merging**